tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,1176 @@
+ * with higher ratios are moved towards the start of the array.
+ */
+static int
+tfw_sched_ratio_srvdata_cmp(const void *lhs, const void *rhs)
+{
+	unsigned int lhs_ratio = ((const TfwRatioSrvData *)lhs)->oratio;
+	unsigned int rhs_ratio = ((const TfwRatioSrvData *)rhs)->oratio;
+
+	if (lhs_ratio > rhs_ratio)
+		return -1;
+	if (lhs_ratio < rhs_ratio)
+		return 1;
+	return 0;
+}
+
+/**
+ * Calculate and set up ratios for each server in the group.
+ *
+ * Return 0 if done with the ratios.
+ * Return a non-zero value if additional actions are needed.
+ */
+static int
+tfw_sched_ratio_calc(TfwRatioPool *rpool, TfwRatio *ratio, size_t *arg_mvidx)
+{
+	size_t si, max_val_idx;
+	unsigned int diff, max_wgt, oratio;
+	unsigned long unit, sum_wgt = 0, sum_ratio = 0;
+	TfwRatioSrvData *srvdata = ratio->srvdata;
+	TfwRatioSchData *schdata = &ratio->schdata;
+
+	BUG_ON(!ratio);
+
+	/*
+	 * Calculate the sum of server's weights in the group. Remember
+	 * the index of server data entry with maximum weight. That same
+	 * entry will also have the maximum ratio. See if all weights in
+	 * the group are the same.
+	 */
+	diff = max_val_idx = 0;
+	for (si = 0; si < rpool->srv_n; ++si) {
+		if (srvdata[max_val_idx].weight < srvdata[si].weight)
+			max_val_idx = si;
+		sum_wgt += srvdata[si].weight;
+		diff |= (srvdata[si].weight != srvdata[0].weight);
+	}
+
+	/* Set up the common part of scheduler data. */
+	schdata->csidx = 0;
+	schdata->riter = 1;
+	schdata->reidx = rpool->srv_n;
+
+	/*
+	 * If all server weights are the same, then there's no need to do
+	 * anything else. Set up all ratios to 1 and be done with it.
+	 */
+	if (!diff) {
+		for (si = 0; si < rpool->srv_n; ++si)
+			srvdata[si].cratio = srvdata[si].oratio = 1;
+		schdata->crsum = schdata->orsum = rpool->srv_n;
+		return 0;
+	}
+
+	/*
+	 * Calculate each server's ratio using a special formula. See
+	 * if all calculated ratios are the same. Set up scheduler data.
+	 */
+	max_wgt = srvdata[max_val_idx].weight;
+	unit = ((max_wgt + rpool->srv_n) * max_wgt) / sum_wgt;
+	for (si = 0; si < rpool->srv_n; ++si) {
+		oratio = (unit * srvdata[si].weight) / max_wgt ? : 1;
+		srvdata[si].cratio = srvdata[si].oratio = oratio;
+		diff |= (oratio != srvdata[0].oratio);
+		sum_ratio += oratio;
+	}
+	schdata->crsum = schdata->orsum = sum_ratio;
+
+	/* Return the index of server data entry with maximum ratio. */
+	*arg_mvidx = max_val_idx;
+
+	return diff;
+}
+
+/*
+ * Calculate and set up ratios for each server in a group based on
+ * weights that are statically defined in the configuration file.
+ */
+static void
+tfw_sched_ratio_calc_static(TfwRatioPool *rpool, TfwRatio *ratio)
+{
+	size_t si, max_val_idx = 0;
+	TfwRatioSrvDesc *srvdesc = rpool->srvdesc;
+	TfwRatioSrvData *srvdata = ratio->srvdata;
+
+	/* Collect server weights from the configuration. */
+	for (si = 0; si < rpool->srv_n; ++si) {
+		srvdata[si].sdidx = si;
+		srvdata[si].weight = srvdesc[si].srv->weight;
+	}
```

Comment:
`tfw_sched_ratio_calc()` is common to static or dynamic weights. However the initial set up is different. Here in the case of static weights the weights are taken from a server instance (structure). In case of dynamic weights the weights are RTT received from APM. How do you propose to eliminate this loop?